RedirectApplication/RedirectApplication/Json/Json.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace RedirectApplication.Json
+{
+    public class JsonWorking
+    {
+        public void Main(string tree)
+        {
+            var reader = new JsonTextReader(new StringReader(tree));
+            PostJson content = JsonSerializer.CreateDefault().Deserialize<PostJson>(reader);
+        }
+    }
+
+    class TreeNodeConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            serializer.Serialize(writer, value);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            var array = JArray.Load(reader);
+            var target = new List<ITreeNode>();
+
+
+            if (array.HasValues)
+            {
+                foreach (var child in array.Children())
+                {
+                    var node = CreateNode(child);
+
+                    serializer.Populate(child.CreateReader(), node);
+                    target.Add(node);
+                }
+            }
+            return target;
+        }
+
+        private ITreeNode CreateNode(JToken obj)
+        {
+            var type = (string)obj["name"];
+
+            switch (type)
+            {
+                case "Composite":
+                    return new Composite();
+                case "ByBrowser":
+                    return new ByBrowser();
+                case "ByLanguage":
+                    return new ByLanguage();
+                case "ByCountry":
+                    return new ByCountry();
+                case "ByIp":
+                    return new ByIp();
+                case "ByOS":
+                    return new ByOS();
+                case "ByDevice":
+                    return new ByDevice();
+                case "ByDate":
+                    return new ByDate();
+                default:
+                    throw new NotSupportedException();
+
+            }
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            return false;
+        }
+    }
+
+    class PostJson
```

Comment:
I propose to move such json definition and `ITreeNode` implemetors classes into a separate files